File: tests/equipment-route.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {readFileSync} from 'node:fs';
import {createHash} from 'node:crypto';
import * as c from '../.test/core.mjs';
import {Vector3,Mesh,MeshBuilder} from '@babylonjs/core';
import {festivalTestScene} from './festival-test-scene.mjs';
import {buildFair} from '../.test/fair-render.mjs';
import {EarlyOcclusion} from '../.test/early-occlusion.mjs';
import {fairOcclusionPoints} from '../.test/fair-occlusion-points.mjs';
import {CONDUCT_POINTS} from '../.test/fair-conduct-data.mjs';

// Unmodified player export from CI34 good-lane prologue, not a manufactured save.
// This file is used ONLY for isolated Node regression tests, never browser input.
const raw=readFileSync(new URL('./fixtures/ci34-companions-v6.json',import.meta.url),'utf8');
const route=JSON.parse(readFileSync(new URL('./equipment-route.json',import.meta.url),'utf8'));
const fresh=()=>c.deserialize(raw);
function advance(s,axis,direction,count){
 const input=[{x:0,z:0},{x:0,z:0}];input[0][axis]=direction;
 for(let n=0;n<count;n++){
  c.step(s,input,1/60);
  assert(s.players.every(p=>c.walkable(p.x,p.z,s.chapter)),'collision must remain authoritative');
 }
}
function leg(s,w,poll,release,guard=true){
 const delta=w.target-s.players[0][w.axis],direction=Math.sign(delta);
 if(Math.abs(delta)<.12)return;
 const budget=Math.ceil((Math.abs(delta)/4+2)*60);let ticks=0;
 while(direction*(w.target-s.players[0][w.axis])>0&&ticks<=budget){advance(s,w.axis,direction,poll);ticks+=poll;}
 assert(ticks<=budget,`blocked ${w.axis} leg to ${w.target} at ${JSON.stringify(s.players[0])}`);
 advance(s,w.axis,direction,release);c.step(s,c.IDLE,1/60);
 if(guard)assert(s.players[0][w.axis]>=w.releaseBand[0]&&s.players[0][w.axis]<=w.releaseBand[1]);
}

test('CI34 companion export is exact, and merchant radius/position match existing core',()=>{
 assert.equal(createHash('sha256').update(raw).digest('hex'),'782d1dc8f23f05ec80c83bd5523c4c79eca40e63443e84500457570e70105db5');
 assert.equal(JSON.parse(raw).version,6);assert.equal(fresh().prologue.stage,'companions');
 assert.equal(route.merchant.x,CONDUCT_POINTS.merchant.x);assert.equal(route.merchant.z,CONDUCT_POINTS.merchant.z);
 assert(readFileSync(new URL('../src/core.ts',import.meta.url),'utf8').includes('distance(s.players[0],CONDUCT_POINTS.merchant)<1.8'));
 assert.equal(route.merchant.radius,1.8);
});

test('CI34 observed candy collision remains solid; longer holding cannot reach old cross-fair target',()=>{
 const s=fresh();
 // Explicit diagnostic fixture reproducing the raw failure observation, not a browser state write.
 Object.assign(s.players[0],{x:5.4,z:-3});
 const before={x:s.players[0].x,z:s.players[0].z};advance(s,'x',1,600);
 assert.deepEqual({x:s.players[0].x,z:s.players[0].z},before);
 assert.equal(c.walkable(5.5,-3,'fair'),false);assert.equal(c.shopAvailable(s),false);
 assert.equal(c.tradeItem(s,'bronze-katana',1,true),false);assert.equal(s.equipment,null);
});

test('old canopy-return route reproduces candy blockage after six extra release ticks',()=>{
 const s=fresh();
 for(const [axis,target] of [['z',-3.4],['x',-8],['z',-4.05]])leg(s,{axis,target},1,0,false);
 leg(s,{axis:'z',target:-3.4},1,6,false);
 assert(s.players[0].z>-3.05);
 assert.throws(()=>leg(s,{axis:'x',target:7.5},6,0,false),/blocked/);
 assert(s.players[0].x<5.46);assert.equal(s.equipment,null);
});

for(const poll of [1,6,12])for(const release of [0,3,6,12])
test(`actual collision-safe canopy/merchant route: polling ${poll}, delayed release ${release}`,()=>{
 const s=fresh(),facts=structuredClone(s.prologue.conduct),flags=structuredClone(s.fair);
 for(const w of route.routes['to-canopy'])leg(s,w,poll,release);
 assert(s.players[0].z>-4.25&&s.players[0].z<=-4);
 // NullEngine geometry check only: the changed stop still uses the real canopy,
 // not a synthetic blocker or a removed browser occlusion assertion.
 const k=festivalTestScene();try{
  const fair=buildFair(k.scene,k.shadow);fair.root.setEnabled(true);
  const mesh=MeshBuilder.CreatePlane('route-player',{width:1.35,height:1.85},k.scene);
  mesh.billboardMode=Mesh.BILLBOARDMODE_ALL;
  mesh.position.copyFrom(new Vector3(s.players[0].x,.14,s.players[0].z).add(k.camera.getDirection(Vector3.Up()).scale(1.85*(62/64-.5))));
  const o=new EarlyOcclusion(k.scene);
  o.update(s.ticks,'fair',k.camera.getDirection(Vector3.Forward()),fairOcclusionPoints([{id:'p0',mesh}]));
  assert(o.inspect().groups.some(g=>g.id==='vq01-festival-cloth-canopy'&&g.blocked));o.dispose();
 }finally{k.dispose();}
 assert.equal(c.shopAvailable(s),false);assert.equal(s.equipment,null);
 // Simulated paused observation has no input; no gameplay clock shortcuts are used.
 for(const w of route.routes['to-merchant'])leg(s,w,poll,release);
 assert.deepEqual(s.prologue.conduct,facts);assert.deepEqual(s.fair,flags);
 assert.equal(s.prologue.stage,'companions');assert.equal(s.joined,false);
 const d=Math.hypot(s.players[0].x-route.merchant.x,s.players[0].z-route.merchant.z);
 assert(d<route.merchant.radius,`merchant distance ${d}`);assert.equal(c.shopAvailable(s),true);
 assert.equal(s.equipment,null,'viewing/reaching the shop must not create gear');
 assert(c.tradeItem(s,'bronze-katana',1,true));assert.equal(s.equipment.gold,250);
 assert(c.equipItem(s,'crono','bronze-katana'));
 const v8=JSON.parse(c.serialize(s));assert.equal(v8.version,8);assert.equal(JSON.parse(v8.adventure).version,6);
 assert.deepEqual(c.deserialize(JSON.stringify(v8)).equipment,s.equipment);
 const loaded=c.deserialize(JSON.stringify(v8));
 for(const w of route.routes['leave-merchant'])leg(loaded,w,poll,release);
 assert.equal(c.shopAvailable(loaded),false);
 for(const w of route.routes['to-gato'])leg(loaded,w,poll,release);
 assert.equal(loaded.mode,'explore');assert.deepEqual(loaded.equipment,s.equipment);
 assert(Math.hypot(loaded.players[0].x-route.gato.x,loaded.players[0].z-route.gato.z)<route.gato.radius);
 assert.match(c.interactFair(loaded,0).title,/岡薩雷斯/);assert.equal(loaded.mode,'battle');
 assert.equal(loaded.players[0].atb,0);assert.equal(loaded.fair.gatoWon,false);

});

test('browser consumes both shared routes and still performs native canopy/pause and merchant checks',()=>{
 const source=readFileSync(new URL('./equipment_browser.py',import.meta.url),'utf8');
 assert(source.includes("walk_equipment_route(page,move,snap,observations,'to-canopy')"));
 assert(source.includes("walk_equipment_route(page,move,snap,observations,'to-merchant')"));
 assert(source.includes("record_festival(page,OUT,'00-cloth-canopy-occlusion',require_blocked=True,pause_probe=True)"));
 assert(source.includes('textContent.includes("裝備買賣")'));
 assert(!source.includes('ci34-companions-v6.json'),'browser must use its own same-run export');
});

File: tests/equipment_route_test.py
"""Unit doubles for route diagnostics only; no browser/device acceptance."""
import copy
import unittest
from unittest.mock import patch
from equipment_route import ROUTE, walk_equipment_route


class Page:
    def __init__(self):
        self.ui = dict(focused='world', paused=False, dialogOpen=False,
                       inventoryOpen=False, hint='', hintHidden=True)
        self.state = dict(chapter='fair', mode='explore', joined=False, ticks=10,
                          prologue={'stage': 'companions'}, players=[{'x': -3.6, 'z': -2.0}])

    def evaluate(self, expression):
        return copy.deepcopy(self.ui)


def snapshot(page):
    return copy.deepcopy(page.state)


class EquipmentRouteTests(unittest.TestCase):
    def setUp(self):
        self.page, self.evidence, self.calls = Page(), [], []

    def move(self, page, axis, target):
        self.calls.append((axis, target))
        page.state['players'][0][axis] = target
        page.state['ticks'] += 1

    def route(self, name='to-canopy', move=None):
        return walk_equipment_route(self.page, move or self.move, snapshot, self.evidence, name)

    def test_shared_waypoints_and_after_release_are_retained_without_changing_snapshot(self):
        trace = self.route()
        self.assertEqual(trace['status'], 'reached')
        self.assertEqual(self.calls, [(w['axis'], w['target']) for w in ROUTE['routes']['to-canopy']])
        self.assertEqual(trace['before']['players'][0], {'x': -3.6, 'z': -2.0})
        self.assertEqual(len(trace['legs']), 3)
        self.assertTrue(all('afterRelease' in leg and 'inputAfter' in leg for leg in trace['legs']))
        self.assertFalse(trace['physicalDevice']); self.assertFalse(trace['artApproved'])

    def test_merchant_turns_south_in_centre_before_crossing_fair(self):
        self.route(); trace = self.route('to-merchant')
        self.assertEqual(self.calls[-3:], [('x', 0), ('z', -5.8), ('x', 7)])
        self.assertLess(trace['distance'], ROUTE['merchant']['radius'])

    def test_unknown_route_never_calls_driver(self):
        with self.assertRaisesRegex(ValueError, 'Unknown'):
            self.route('shortcut')
        self.assertEqual(self.calls, []); self.assertEqual(self.evidence, [])

    def test_invalid_initial_context_never_sends_movement(self):
        changes = [('chapter', 'lab'), ('mode', 'battle'), ('joined', True), ('ticks', float('nan'))]
        for key, value in changes:
            with self.subTest(key=key):
                self.setUp(); self.page.state[key] = value
                with self.assertRaises(AssertionError): self.route()
                self.assertEqual(self.calls, []); self.assertEqual(self.evidence[0]['status'], 'failed')
        self.setUp(); self.page.state['prologue']['stage'] = 'collision'
        with self.assertRaises(AssertionError): self.route()
        self.assertEqual(self.calls, [])

    def test_modal_or_lost_focus_stops_before_first_key(self):
        for key, value in [('paused', True), ('dialogOpen', True), ('inventoryOpen', True), ('focused', 'save-file')]:
            with self.subTest(key=key):
                self.setUp(); self.page.ui[key] = value
                with self.assertRaises(AssertionError): self.route()
                self.assertEqual(self.calls, [])

    def test_out_of_aisle_after_keyup_does_not_start_next_leg(self):
        def overshot(page, axis, target):
            self.move(page, axis, target); page.state['players'][0][axis] = -4.3
        with self.assertRaises(AssertionError): self.route(move=overshot)
        self.assertEqual(len(self.calls), 1)
        self.assertEqual(self.evidence[0]['legs'][0]['afterRelease']['players'][0]['z'], -4.3)
        self.assertEqual(self.evidence[0]['status'], 'failed')

    def test_driver_failure_is_original_and_retains_post_release_diagnostics(self):
        original = RuntimeError('original movement budget exhausted')
        def failed(page, axis, target):
            self.move(page, axis, target); raise original
        with self.assertRaises(RuntimeError) as raised: self.route(move=failed)
        self.assertIs(raised.exception, original)
        self.assertEqual(len(self.calls), 1)
        self.assertIn('afterRelease', self.evidence[0]['legs'][0])

    def test_diagnostic_error_does_not_replace_movement_failure(self):
        original = RuntimeError('original movement failure')
        def failed(page, axis, target):
            self.move(page, axis, target)
            page.evaluate = lambda expression: (_ for _ in ()).throw(ValueError('lost page'))
            raise original
        with self.assertRaises(RuntimeError) as raised: self.route(move=failed)
        self.assertIs(raised.exception, original)
        self.assertIn('captureError', self.evidence[0]['legs'][0])

    def test_context_change_after_keyup_stops_next_leg(self):
        for change in ['focus', 'clock', 'stage', 'joined', 'coordinate']:
            with self.subTest(change=change):
                self.setUp()
                def changed(page, axis, target):
                    self.move(page, axis, target)
                    if change == 'focus': page.ui['focused'] = 'import'
                    elif change == 'clock': page.state['ticks'] = 0
                    elif change == 'stage': page.state['prologue']['stage'] = 'fair'
                    elif change == 'joined': page.state['joined'] = True
                    else: page.state['players'][0]['x'] = float('nan')
                with self.assertRaises(AssertionError): self.route(move=changed)
                self.assertEqual(len(self.calls), 1)

    def test_in_band_endpoint_still_needs_real_merchant_distance(self):
        self.route()
        def far_corner(page, axis, target):
            self.move(page, axis, target)
            if axis == 'x' and target == 7:
                page.state['players'][0] = {'x': 8.6, 'z': -7.4}
        with patch.dict(ROUTE, merchant={**ROUTE['merchant'], 'radius': 1.0}):
            with self.assertRaises(AssertionError): self.route('to-merchant', far_corner)
        self.assertEqual(self.evidence[-1]['status'], 'failed')
        self.assertGreater(self.evidence[-1]['distance'], 1.0)


if __name__ == '__main__':
    unittest.main()
